rfc/text/advanced/ap_authentication_scram.md

```diff
@@ -0,0 +1,387 @@
+
+2. Hashing based on the weaker SHA-1 specified in [RFC5802](https://tools.ietf.org/html/rfc5802) is intentionally not supported by WAMP-SCRAM, in favor of the stronger SHA-256 specified in [RFC7677](https://tools.ietf.org/html/rfc7677).
+
+3. The [Argon2](https://datatracker.ietf.org/doc/draft-irtf-cfrg-argon2) key derivation function MAY be used instead of PBKDF2.
+
+4. Nonces are required to be base64-encoded, which is stricter than the `printable` syntax specification of [RFC5802](https://tools.ietf.org/html/rfc5802).
+
+5. The `"y"` channel binding flag is not used as there is currently no standard way for WAMP routers to announce channel binding capabilities.
+
+6. The use of `authzid` for user impersonation is not supported.
+
+##### authmethod Type String
+
+`"wamp-scram"` SHALL be used as the `authmethod` type string for WAMP-SCRAM authentication. Announcement by routers of WAMP-SCRAM support is outside the scope of this document.
+
+##### Conventions
+
+###### Base64 encoding
+
+Base64 encoding of octet strings is restricted to canonical form with no whitespace, as defined in [RFC4648](https://tools.ietf.org/html/rfc4648) (_The Base16, Base32, and Base64 Data Encodings_).
+
+###### Nonces
+
+In WAMP-SCRAM, a _nonce_ (number used once) is a base64-encoded sequence of random octets. It SHOULD be of sufficient length to make a replay attack unfeasible.
+
+See [RFC4086](https://tools.ietf.org/html/rfc4086) (_Randomness Requirements for Security_) for best practices concerning randomness.
+
+###### Salts
+
+A _salt_ is a base64-encoded sequence of random octets.
+
+To prevent rainbow table attacks in the event of database theft, the salt MUST be generated randomly by the server **for each user**. The random salt is stored with each user record in the authentication database.
+
+###### Username/Password String Normalization
+
+Username and password strings SHALL be normalized according to the _SASLprep_ profile described in [RFC4013](https://tools.ietf.org/html/rfc4013), using the _stringprep_ algorithm described in [RFC3454](https://tools.ietf.org/html/rfc3454).
+
+While SASLprep preserves the case of usernames, the server MAY choose to perform case insensitve comparisons when searching for a username in the authentication database.
+
+
+##### Channel Bindings
+
+_Channel binding_ is a feature that allows higher layers to establish that the other end of an underlying secure channel is bound to its higher layer counterpart. See [RFC5056](https://tools.ietf.org/html/rfc5056) (_On the Use of Channel Bindings_) for an in-depth discussion.
+
+[RFC5929](https://tools.ietf.org/html/rfc5929) defines binding types for use with TLS transports, of which `"tls-server-end-point"` may be used with SCRAM. Note that [RFC7677, section 1](https://tools.ietf.org/html/rfc7677#section-1), states that TLS did not have the expected properties for `"tls-unique"` to be secure.
```

Comment:
Mmh. "did not have the expected properties" for what? =) seriously, here is how I read it. the RFC7627 introduces a "Extended Master Secret Extension" which when used makes the master secret, and thereby the session hash, cover _also_ the peer certificate. and then, SASL would be secure when using tls-unique channel binding. IOW: we would say SHOULD use channel binding and SHOULD use extended master secret extension. With tls-server-end-point: as I understand that, the channel is "only" bound to the public bits of the peer certificate in this case - or the private bits too?